### PR TITLE
RUMM-182 Sync logs time with server time

### PR DESCRIPTION
**Pull request by buranmert**
### What and why?

• SDK needs to send current date in miliseconds in order for the backend to be able to tell time difference between the device and actual time. 
• This is especially essential for tracing among different systems.

### How?

1. `DataUploadURL` can change query of its URL dynamically
2. `URLQueryProvider` is made `enum` so that we limit the number of cases to test

### Review checklist

- [x] Feature or bugfix MUST have appropriate tests (unit, integration)
- [x] Make sure each commit and the PR mention the Issue number or JIRA reference

### CODE REVIEW

File: Sources/Datadog/Core/Upload/DataUploader.swift [hunk 1]
```diff
@@ -6,32 +6,52 @@
...
-    let url: URL
+/// Creates URL and adds query items before providing them
+internal class UploadURLProvider {
+    private let rawURL: URL
+    private let dateProvider: DateProvider
 
-    init(endpointURL: String, clientToken: String) throws {
+    private var queryItems: [URLQueryItem] {
+        // batch_time
+        let currentDateInMs = dateProvider.currentDate().timeIntervalSinceReferenceDate
```
Comment (ncreated) [inline Sources/Datadog/Core/Upload/DataUploader.swift:16]: 1/ Should use `. timeIntervalSinceReferenceDate()`? AFAIK it's the interval since 1st January of 2000. Shouldn't it be timestamp since 1970? 
2/ Also, this is in seconds, does backend require seconds or milliseconds / microseconds?

File: Sources/Datadog/Core/Upload/DataUploader.swift [hunk 2]
```diff
@@ -6,32 +6,52 @@
...
+        let ddSourceQueryItem = URLQueryItem(name: "ddsource", value: "mobile")
+
+        return [ddSourceQueryItem, batchTimeQueryItem]
+    }
+
+    var url: URL {
+        var urlComponents = URLComponents(url: rawURL, resolvingAgainstBaseURL: false)
+        urlComponents?.percentEncodedQueryItems = queryItems
+
+        // TODO: RUMM-182 how to handle nil value?
```
Comment (ncreated) [inline Sources/Datadog/Core/Upload/DataUploader.swift:28]: IMO best we can do is to `guard let` and print appropriate errors on `sdkLogger` and `userLogger`.

File: Sources/Datadog/Datadog.swift [hunk 3]
```diff
@@ -103,14 +103,14 @@ public class Datadog {
...
-        logsUploadURL: DataUploadURL,
+        logsUploadURL: UploadURLProvider,
         dateProvider: DateProvider,
         userInfoProvider: UserInfoProvider,
         networkConnectionInfoProvider: NetworkConnectionInfoProviderType,
         carrierInfoProvider: CarrierInfoProviderType?
     ) throws {
         let logsPersistenceStrategy: LogsPersistenceStrategy = try .defalut(using: dateProvider)
-        let logsUploadStrategy: LogsUploadStrategy = .defalut(
+        let logsUploadStrategy: LogsUploadStrategy = .default(
```
Comment (ncreated) [inline Sources/Datadog/Datadog.swift:113]: ⭐️ 🏆

File: Sources/Datadog/Core/Upload/DataUploader.swift (hunk 1 above)
Comment (buranmert) [reply Sources/Datadog/Core/Upload/DataUploader.swift:16]: i verified with https://github.com/DataDog/dd-sdk-android/pull/181/files
they are using [`currentTimeMillis` from Java](https://docs.oracle.com/javase/7/docs/api/java/lang/System.html#currentTimeMillis())

https://github.com/DataDog/dd-sdk-android/blob/master/dd-sdk-android/src/main/kotlin/com/datadog/android/core/internal/time/DatadogTimeProvider.kt

changed ✅

File: Sources/Datadog/DatadogConfiguration.swift [hunk 4]
```diff
@@ -62,10 +63,19 @@ extension Datadog {
...
+            internal func set(dateProvider: DateProvider) -> Builder {
+                self.dateProvider = dateProvider
+                return self
+            }
+
             /// Builds `Datadog.Configuration` object.
             public func build() -> Configuration {
                 return Configuration(
-                    logsUploadURL: try? DataUploadURL(endpointURL: logsEndpoint.url, clientToken: clientToken)
+                    logsUploadURLProvider: try? UploadURLProvider(
```
Comment (ncreated) [inline Sources/Datadog/DatadogConfiguration.swift:74]: Hmm, IMO we have two problems in `DatadogConfiguration`:
* 1/ The `set(dateProvider: DateProvider) -> Builder` is there only for unit tests;
* 2/ We don't leverage `throwable` initializer of `UploadURLProvider`, so the `ProgrammerErrors` thrown from there are not printed to the user.

IMO we can fix both by simply changing from:
```swift
struct DatadogConfiguration {
   internal let logsUploadURLProvider: UploadURLProvider?
   // ...
}
```
to:
```swift
struct DatadogConfiguration {
   internal let endpointURL: URL
   internal let clientToken: String
   // ...
}
```
and delegating `UploadURLProvider.init(...)` to `Datadog`.

`Datadog` consumes `DatadogConfiguration` in:
```swift
// Datadog.swift
private static func initializeOrThrow(appContext: AppContext, configuration: Configuration) throws 
```
so we might `rethrow` the error from `UploadURLProvider` and it will be printed for the user.

This will solve both 1/ and 2/. Also, this will print `ProgrammerError` to the user if `UploadURLProvider` cannot be constructed, which we can unit test in `DatadogTests`.

Comment (buranmert) [reply Sources/Datadog/DatadogConfiguration.swift:74]: super idea 🙌 
i didn't like what i did neither 😕 
i'm changing ♻️

File: Sources/Datadog/Utils/SwiftExtensions.swift [hunk 5]
```diff
@@ -13,3 +13,11 @@ extension Optional {
         }
     }
 }
+
+extension Date {
+    // NOTE: RUMM-182 counterpart of currentTimeMillis in Java
+    // https://docs.oracle.com/javase/7/docs/api/java/lang/System.html#currentTimeMillis()
+    var currentTimeMillis: UInt {
+        return UInt(self.timeIntervalSince1970 * 1_000)
```
[the remark this answers is not in the document]
Comment (mariusc83) [reply Sources/Datadog/Utils/SwiftExtensions.swift:21]: why  "<" than UInt64.max and not "<=" ?

[the remark this answers is not in the document]
Comment (ncreated) [reply Sources/Datadog/Utils/SwiftExtensions.swift:21]: Hm, no reason I think, we just overlooked this 🤔.

File: Tests/DatadogTests/Datadog/Utils/FoundationExtensionTests.swift [hunk 6]
```diff
@@ -0,0 +1,22 @@
+/*
+ * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2019-2020 Datadog, Inc.
+ */
+
+import XCTest
+@testable import Datadog
+
+@available(iOS 13, *)
```
[the remark this answers is not in the document]
Comment (buranmert) [reply Tests/DatadogTests/Datadog/Utils/FoundationExtensionTests.swift:10]: because `advanced(:)` is an iOS 13 method.
i just found out that i could do simple `+, -` operations on `Date` in earlier versions

done ✅

File: Sources/Datadog/Utils/SwiftExtensions.swift (hunk 5 above)
[the remark this answers is not in the document]
Comment (buranmert) [reply Sources/Datadog/Utils/SwiftExtensions.swift:21]: this solution does not cover all the cases.

`fileCreationDate.timeIntervalSinceReferenceDate * 1_000` may overflow as well

[the remark this answers is not in the document]
Comment (ncreated) [reply Sources/Datadog/Utils/SwiftExtensions.swift:21]: @bmermet yes, and that's exactly why we check for overflow in `FilesOrchestrator`. We can reuse that code and create extension for safe conversion:

Screenshot 2020-04-07 at 18 22 19

[the remark this answers is not in the document]
Comment (buranmert) [reply Sources/Datadog/Utils/SwiftExtensions.swift:21]: This code above crashes if `double` is close to `Double.greatestFiniteMagnitude`

That one below seems like a better solution
```swift
extension FixedWidthInteger {
    init(fromDoubleWithReportingOverflow double: Double) throws {
        guard let converted = Self(exactly: double.rounded()) else {
            throw FixedWidthIntegerError.overflow(overflowingDouble: double)
        }
        self = converted
    }
}

enum FixedWidthIntegerError: Error {
    case overflow(overflowingDouble: Double)
}
```

1. this extension works for all integer types
2. it doesn't contain confusing logic (eg: <UInt64.min, >=UInt32.max, etc.)
3. `init(exactly:)` is a failable initializer, so it doesn't throw or crash

[the remark this answers is not in the document]
Comment (ncreated) [reply Sources/Datadog/Utils/SwiftExtensions.swift:21]: What does it mean "if `double` is close to `Double.greatestFiniteMagnitude`" 🤔?

[the remark this answers is not in the document]
Comment (buranmert) [reply Sources/Datadog/Utils/SwiftExtensions.swift:21]: I haven't calculated the exact range yet but if `double` is between little more than `0.9999999999999999` and `1.0` times of `Double.greatestFiniteMagnitude`, it **crashes**.

[`init(exactly:)`](https://github.com/apple/swift/blob/master/stdlib/public/core/Integers.swift#L3080) uses the same method with [`init(:)`](https://github.com/apple/swift/blob/master/stdlib/public/core/Integers.swift#L3054) internally

[the remark this answers is not in the document]
Comment (ncreated) [reply Sources/Datadog/Utils/SwiftExtensions.swift:21]: `Double.greatestFiniteMagnitude` is an abstraction over "infinity", so it shouldn't be used for arithmetic operations - this is IMHO asking for trouble 💥.

Whatever works is OK for me 👍. However, the plain `double >= UInt64.min && double  <= UInt64.max` is way more readable (also intuitive and proven on production) than `FixedWidthInteger(exactly: double.rounded())`.

[the remark this answers is not in the document]
Comment (buranmert) [reply Sources/Datadog/Utils/SwiftExtensions.swift:21]: Yes, I agree it is not intuitive, unfortunately it took me quite a while to find that method.

Here is my pros&cons:

### Pros:

1. We don't need to do sanity-check (ie: `double >= UInt64.min`), Swift does that for us 🎉 
    * It has less code and doesn't add extra logic
2. It never crashes, it works at even corner cases such as `Double.greatestFiniteMagnitude`
3. It has the same level of correctness with `init(:)` as they use the same internal method, the only difference is that this doesn't do `fatalError()`
    * I left URLs to their source code in the comment above ☝️ 

### Cons:

1. `init(exactly:)` is not a method name that you think of first when you are dealing with overflows
    * however, if we keep it in an extension and provide a better method name, it should be fine

[the remark this answers is not in the document]
Comment (mariusc83) [reply Sources/Datadog/Utils/SwiftExtensions.swift:21]: IMO ....I don't think you need any type safety check here as you are trying to convert a Date Double which will always be positive and smaller than Uint64.max (because Double is using his MSB for sign). If you are sure that this Date can never be negative you can just do the conversion directly ....if not you can just check if is ">=0" only.

[the remark this answers is not in the document]
Comment (mariusc83) [reply Sources/Datadog/Utils/SwiftExtensions.swift:21]: Ok nevermind what I said there....I definitely forgot how a Double is being represented. You definitely need that check there.

[the remark this answers is not in the document]
Comment (buranmert) [reply Sources/Datadog/Utils/SwiftExtensions.swift:21]: ### Safety check

_TL;DR: If we do boundry/safety check here, it will be nothing more than duplicating Swift's internal checks._

Swift does the exact same safety checks internally anyway.
The difference is, in case of failure:
1. `init(:)` calls `fatalError` (this is different than `error` or `exception`, the app just crashes) ❌ 
2. `init(exactly:)` returns `nil` ✅

[the remark this answers is not in the document]
Comment (mariusc83) [reply Sources/Datadog/Utils/SwiftExtensions.swift:21]: In this case I would go with the ">=,<" solution as it is more readable for me.

File: Tests/DatadogTests/DatadogObjc/DDConfigurationTests.swift [hunk 7]
```diff
@@ -1,36 +0,0 @@
...
-    // MARK: - Customizing logs endpoint
-
-    private let swiftEndpoints: [Datadog.Configuration.LogsEndpoint] = [
-        .eu, .us, .custom(url: "https://api.example.com/v1/logs")
-    ]
-    private let objcEndpoints: [DDLogsEndpoint] = [
-        .eu(), .us(), .custom(url: "https://api.example.com/v1/logs")
-    ]
-
-    func testItForwardsLogsEndpointToSwift() {
```
Comment (ncreated) [inline Tests/DatadogTests/DatadogObjc/DDConfigurationTests.swift:22]: This was testing if `DDConfiguration` initialization leads to expected `Datadog.Configuration`. With this removal we lost coverage on `DDLogsEndpoint` → `Datadog.Configuration.LogsEndpoint` mapping:

Screenshot 2020-04-08 at 17 05 17

File: Sources/Datadog/Datadog.swift [hunk 8]
```diff
@@ -88,12 +88,16 @@ public class Datadog {
...
             throw ProgrammerError(description: "SDK is already initialized.")
         }
-        guard let logsUploadURL = configuration.logsUploadURL else {
-            throw ProgrammerError(description: "SDK configuration is invalid - check `logsEndpoint` and (or) `clientToken`.")
+        guard let logsUploadURLProvider = try? UploadURLProvider(
+            endpointURL: configuration.logsEndpoint.url,
+            clientToken: configuration.clientToken,
+            dateProvider: SystemDateProvider()
+            ) else {
+                throw ProgrammerError(description: "SDK configuration is invalid - check `logsEndpoint` and (or) `clientToken`.")
```
Comment (ncreated) [inline Sources/Datadog/Datadog.swift:96]: Let's just re-throwing the existing errors. Now we capture them with `try?` and erase the error information by throwing another `ProgrammerError`. This would be easier and nicer and shorter:
```swift
let logsUploadURLProvider = try UploadURLProvider(
   endpointURL: configuration.logsEndpoint.url,
   clientToken: configuration.clientToken,
   dateProvider: SystemDateProvider()
)
```

The two distinct errors are already thrown from `UploadURLProvider.init`:

```swift
// UploadURLProvider

init(endpointURL: String, clientToken: String, dateProvider: DateProvider) throws {
   guard !endpointURL.isEmpty, let endpointURL = URL(string: endpointURL) else {
      throw ProgrammerError(description: "`endpointURL` cannot be empty.")
   }
   guard !clientToken.isEmpty else {
      throw ProgrammerError(description: "`clientToken` cannot be empty.")
   }

   // ...
}
```

File: Sources/Datadog/Utils/SwiftExtensions.swift [hunk 9]
```diff
@@ -6,10 +6,52 @@
...
         }
     }
 }
+
+// MARK: - Date
+
+extension Date {
+    // NOTE: RUMM-182 counterpart of currentTimeMillis in Java
+    // https://docs.oracle.com/javase/7/docs/api/java/lang/System.html#currentTimeMillis()
+    var currentTimeMillis: UInt {
```
[the remark this answers is not in the document]
Comment (buranmert) [reply Sources/Datadog/Utils/SwiftExtensions.swift:24]: because it overflows. so we should explicitly use `UInt64`?

File: Sources/Datadog/Datadog.swift (hunk 8 above)
Comment (buranmert) [reply Sources/Datadog/Datadog.swift:96]: good catch 💯 i missed that while copy&pasting

File: Tests/DatadogTests/DatadogObjc/DDConfigurationTests.swift (hunk 7 above)
Comment (buranmert) [reply Tests/DatadogTests/DatadogObjc/DDConfigurationTests.swift:22]: when i `DDDatadog.initialize` to implicitly create `UploadURLProvider`, i don't see a way to access to it.

the path seems like the following
`Datadog->instance->logsUploadStrategy->uploadWorker->(private) dataUploader->(private) urlProvider`

do we have a better idea than making them `internal` just for tests?

File: Sources/Datadog/Utils/SwiftExtensions.swift (hunk 9 above)
[the remark this answers is not in the document]
Comment (ncreated) [reply Sources/Datadog/Utils/SwiftExtensions.swift:24]: > because it overflows. so we should explicitly use UInt64?

Yes, it's the only way to represent current timestamp on `32-bit` devices. This will be less performant as not optimised in CPU, but we have no other way.

File: Tests/DatadogTests/DatadogObjc/DDConfigurationTests.swift (hunk 7 above)
Comment (ncreated) [reply Tests/DatadogTests/DatadogObjc/DDConfigurationTests.swift:22]: `DDConfigurationTests` was only asserting if given `DDConfiguration` builds expected Swift `Datadog.Configuration`, so it doesn't configure nor call `Datadog.initialize` under the hood.

As in this PR we change `Datadog.Configuration` to be basic `struct` with `clientToken` and `logsUploadEndpoint` being the only properties, this should be enough to have 100% coverage on `DDConfiguration`:

```swift
import XCTest
@testable import Datadog
@testable import DatadogObjc

/// This tests verify that objc-compatible `DatadogObjc` wrapper properly interacts with`Datadog` public API (swift).
class DDConfigurationTests: XCTestCase {
    func testItFowardsInitializationToSwift() {
        let objcBuilder = DDConfiguration.builder(clientToken: "abc-123")

        let swiftConfigurationDefault = objcBuilder.build().sdkConfiguration
        XCTAssertEqual(swiftConfigurationDefault.clientToken, "abc-123")
        XCTAssertEqual(swiftConfigurationDefault.logsEndpoint, .us)

        objcBuilder.set(endpoint: .eu())
        let swiftConfigurationEU = objcBuilder.build().sdkConfiguration
        XCTAssertEqual(swiftConfigurationEU.logsEndpoint, .eu)

        objcBuilder.set(endpoint: .us())
        let swiftConfigurationUS = objcBuilder.build().sdkConfiguration
        XCTAssertEqual(swiftConfigurationUS.logsEndpoint, .us)

        objcBuilder.set(endpoint: .custom(url: "https://api.example.com/v1/logs"))
        let swiftConfigurationCustom = objcBuilder.build().sdkConfiguration
        XCTAssertEqual(swiftConfigurationCustom.logsEndpoint, .custom(url: "https://api.example.com/v1/logs"))
    }
}

extension Datadog.Configuration.LogsEndpoint: Equatable {
    public static func == (_ lhs: Datadog.Configuration.LogsEndpoint, _ rhs: Datadog.Configuration.LogsEndpoint) -> Bool {
        switch (lhs, rhs) {
        case (.us, .us): return true
        case (.eu, .eu): return true
        case let (.custom(lhsURL), .custom(rhsURL)): return lhsURL == rhsURL
        default: return false
        }
    }
}
```

💡 Please note that I hand-crafted `Equatable` conformance in this test (vs marking it conform to `Equatable` in `Datadog.framework`) as we rather don't want the `LogsEndpoint` to be `Equatable` in public interface as that would require backward compatibility consideration with each change.

Comment (buranmert) [reply Tests/DatadogTests/DatadogObjc/DDConfigurationTests.swift:22]: copied&pasted ✅ 
but let's reconsider this builder for `v2.0` please ♻️

File: Tests/DatadogTests/Datadog/Utils/SwiftExtensionsTests.swift [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+    func testCurrentMillis() {
+        let date15Dec2019 = Date.mockDecember15th2019At10AMUTC()
+        XCTAssertEqual(date15Dec2019.currentTimeMillis, 1_576_404_000_000)
+
+        let dateAdvanced = date15Dec2019 + 9.999
+        XCTAssertEqual(dateAdvanced.currentTimeMillis, 1_576_404_009_999)
+
+        let dateAgo = date15Dec2019 - 0.001
+        XCTAssertEqual(dateAgo.currentTimeMillis, 1_576_403_999_999)
+    }
```
Comment (ncreated) [inline Tests/DatadogTests/Datadog/Utils/SwiftExtensionsTests.swift:20]: Let's also include test for the code branch which captures overflow, this can be easily done by using "reference date" as relative base:

```suggestion
    }

    func testUInt64Overflow() {
        let dateWithOverflow = Date(timeIntervalSinceReferenceDate: .greatestFiniteMagnitude)
        XCTAssertEqual(dateWithOverflow.currentTimeMillis, UInt64.max)
    }
```

File: Tests/DatadogTests/Datadog/DatadogConfigurationTests.swift [hunk 11]
```diff
@@ -1,50 +0,0 @@
...
- * Copyright 2019-2020 Datadog, Inc.
- */
-
-import XCTest
-@testable import Datadog
-
-class DatadogConfigurationTests: XCTestCase {
-    private typealias Configuration = Datadog.Configuration
-
-    func testDefaultConfiguration() {
```
Comment (ncreated) [inline Tests/DatadogTests/Datadog/DatadogConfigurationTests.swift:13]: We no longer test `Configuration.Builder` behaviours. In fact, when we switch default endpoint from `.us` to `.eu`, none of Swift unit tests do fail.

This will test builder logic:
```swift
import XCTest
@testable import Datadog

class DatadogConfigurationTests: XCTestCase {
    private typealias Configuration = Datadog.Configuration

    func testDefaultConfiguration() {
        let defaultConfiguration = Configuration.builderUsing(clientToken: "abcd").build()
        XCTAssertEqual(defaultConfiguration.clientToken, "abcd")
        XCTAssertEqual(defaultConfiguration.logsEndpoint.url, "https://mobile-http-intake.logs.datadoghq.com/v1/input/")
    }

    // MARK: - Logs endpoint
    func testUSLogsEndpoint() {
        let configuration = Configuration.builderUsing(clientToken: .mockAny()).set(logsEndpoint: .us).build()
        XCTAssertEqual(configuration.logsEndpoint.url, "https://mobile-http-intake.logs.datadoghq.com/v1/input/")
    }

    func testEULogsEndpoint() {
        let configuration = Configuration.builderUsing(clientToken: .mockAny()).set(logsEndpoint: .eu).build()
        XCTAssertEqual(configuration.logsEndpoint.url, "https://mobile-http-intake.logs.datadoghq.eu/v1/input/")
    }

    func testCustomLogsEndpoint() {
        let configuration = Configuration.builderUsing(clientToken: .mockAny())
            .set(logsEndpoint: .custom(url: "https://api.example.com/v1/logs/"))
            .build()
        XCTAssertEqual(configuration.logsEndpoint.url, "https://api.example.com/v1/logs/")
    }
}
```

File: Tests/DatadogTests/Datadog/Core/Upload/DataUploaderTests.swift [hunk 12]
```diff
@@ -7,26 +7,62 @@
...
+
+    func testCustomLogsEndpoint() {
+        let urlCustom = Datadog.Configuration.LogsEndpoint.custom(url: "foo.bar").url
+        let urlProvider = try! UploadURLProvider(
+            endpointURL: urlCustom,
+            clientToken: "abc",
+            dateProvider: dateProvider
+        )
+        XCTAssertEqual(urlProvider.url, URL(string: "foo.bar/abc?ddsource=mobile&batch_time=1576404000000"))
+    }
```
Comment (ncreated) [inline Tests/DatadogTests/Datadog/Core/Upload/DataUploaderTests.swift:65]: Those tests do exactly the same assertion as `func testItBuildsValidURL()` - _"When a given URL string is provided, `UploadURLProvider` decorates it with expected URL queries"_. Asserting values of the URL for different endpoints (`.us`, `.eu`, `.custom(...)`) is a better fit for configuration builder tests (`DatadogConfigurationTests`) as we also need to assert the defaults it produces (when no endpoint is set explicitly).

In other words:
* in `DataUploaderTests` we can test: 
   * `.us` endpoint string
   * `.eu` endpoint string
   * `.custom` endpoint string
* in `DatadogConfigurationTests ` we could test: 
   * `.us` endpoint string
   * `.eu` endpoint string
   * `.custom` endpoint string
   * default endpoint if not explicitly set by the user 👈 


```suggestion
```

File: Tests/DatadogTests/Datadog/Utils/SwiftExtensionsTests.swift (hunk 10 above)
Comment (buranmert) [reply Tests/DatadogTests/Datadog/Utils/SwiftExtensionsTests.swift:20]: good idea 💡 
i just noticed i forgot to add tests for `init(withReportingOverflow:)` too

File: Tests/DatadogTests/Datadog/Core/Upload/DataUploaderTests.swift (hunk 12 above)
Comment (buranmert) [reply Tests/DatadogTests/Datadog/Core/Upload/DataUploaderTests.swift]: done ✅

File: Tests/DatadogTests/Datadog/DatadogConfigurationTests.swift (hunk 11 above)
Comment (buranmert) [reply Tests/DatadogTests/Datadog/DatadogConfigurationTests.swift:13]: copied&pasted ✅
